Use IGRAB-compatible names for graphics chunks

sulix opened the pull request:
Bio Menace's graphics were originally imported using a tool called IGRAB, which generates the GRAPHBM1.H and GRAPHBM1.EQU files with the chunk numbers, and other metadata.

IGRAB, however, has a number of rules when generating these names (see FINISH.C):
- Bitmaps have 'PIC' appended to the name.
- Masked bitmaps have 'PICM' appended to the name.
- Sprites have 'SPR' appended to the name.
- (Other chunks have no naming restrictions)

The Bio Menace decomp uses a handwritten GRAPHBM1.{H,EQU} file, as the originals are not available (and neither is the IGRAB .I script). However, the names used do not conform to the original IGRAB rules, so it would not be possible to reconstruct such a file, or to use other tools which expect such formatting, to create compatible headers.

In particular:
- The 'ceilwalker' graphics chunks are missing a 'SPR' suffix.
- Where a 'SPR' suffix is used, it is separated with an underscore, making it '_SPR'. This is not a problem, but does look a bit weird (IGRAB must be given a name with a trailing underscore).

This change mechanically adds 'SPR' as a suffix where it was not present, and removes the '_' from '_SPR' where used. This is still a bit ugly, but it is representable.

I've tested this with headers generated from my own in-progress [idGrab](https://github.com/sulix/idgrab) tool, which can generate IGRAB-style headers, and the game builds and runs fine.

As an example, idGrab can use [this script](https://davidgow.net/stuff/bmenace1.idgrab) to generate:
- [GRAPHBM1.H](https://davidgow.net/stuff/GRAPHBM1.H)
- [GRAPHBM1.EQU](https://davidgow.net/stuff/GRAPHBM1.EQU)


What do you think?
Personally, I'd like to at least give CEILWALKER the SPR ending, but could learn to live with the underscores (I do agree they look nicer in the manually-made headers.)

lethal-guitar commented:
Hi, and many thanks for the contribution, appreciate it!

Also, great work on the `idGrab` tool!

I like the idea of being able to auto-generate these files, so I'm generally in favor of this change. I do think readability suffers a fair bit without the underscores though, at least to my eyes - it makes the suffixes I use for numbering and left/right variations blend in more with the `SPR` suffix, and thus harder to identify.

Now, based on the general style of ID's source code that has been published, my guess is that the original version most likely used no underscores at all in the sprite/graphic names, similar to most variable names etc. So for the most authentic recreation, that would likely be the way to go. But when it comes to decompilations, I generally prefer to improve readability where I can, even if that's likely to be less accurate to the original source code, as long as it doesn't impact the accuracy of the binary files produced from the code.

So my preference would be to only add the missing `SPR` suffixes. I can see that it makes the names in the source script a little awkward, but that seems like an acceptable tradeoff to me. Especially since many of the sprite names are directly referenced in non-generated code, like the various state definitions.

Let me know what you think!

sulix commented:
Sounds like a plan!

Should we also add an underscore to the bitmaps / masked bitmaps? (e.g. ``CP_MAINMENUPIC`` → ``CP_MAINMENU_PIC``)

That would give us some more consistency ([example GRAPHBM1.H](https://davidgow.net/stuff/GRAPHBM1_H_underscores.txt)), but make those bitmap names less authentic to the original. (``H_HELPPIC``, for example, matches [the original enum from Keen 4](https://davidgow.net/stuff/k4_chunkenum.txt), but ``H_HELP_PIC`` wouldn't).

lethal-guitar commented:
> Should we also add an underscore to the bitmaps / masked bitmaps?

Ah good point, yeah I think that's a good idea. I started out with the Keen 4 header, which is why many of the bitmaps have identical names. I then started using underscores when adding the sprite names, but I didn't go back and change the existing names. Would definitely be nice to be consistent.

sulix committed:
Use IGRAB-compatible names for graphics chunks

Bio Menace's graphics were originally imported using a tool called
IGRAB, which generates the GRAPHBM1.H and GRAPHBM1.EQU files with the
chunk numbers, and other metadata.

IGRAB, however, has a number of rules when generating these names:
- Bitmaps have 'PIC' appended to the name.
- Masked bitmaps have 'PICM' appended to the name.
- Sprites have 'SPR' appended to the name.
- (Other chunks have no naming restrictions)

The Bio Menace decomp uses a handwritten GRAPHBM1.{H,EQU} file, as the
originals are not available (and neither is the IGRAB .I script).
However, the names used do not conform to the original IGRAB rules, so
it would not be possible to reconstruct such a file, or to use other
tools which expect such formatting, to create compatible headers.

In particular:
- The 'ceilwalker' graphics chunks are missing a 'SPR' suffix.
- Where a 'SPR' suffix is used, it is often (but not always) separated
  with an underscore, making it '_SPR'. This is not a problem, but does
  look a bit weird (IGRAB must be given a name with a trailing
  underscore).
- The PIC and PICM suffixes are not separated by an underscore, which is
  inconsistent.

This change mechanically adds '_SPR' as a suffix where it was not
present, and adds an underscore where one was missing (for PIC, PICM,
and a few sprites like SCOREBOXSPR). This provides consistency, and is
technically compatible with IGRAB, so long as all of the bitmap, masked
bitmap, and sprite names end in a trailing underscore.

I've tested this with headers generated from my own in-progress 'idGrab'
tool, which can generate IGRAB-style headers, (and can automatically add
an underscore as a separator), and the game builds and runs fine.

sulix commented:
Sounds good — I've pushed a new version which adds the underscore consistently, and it seems to work fine. There were a few more sprites which were missing the underscore:
- ``HANDCURSOR``
- ``PLATFORM``
- ``PUSHBLOCK``
- ``CRUSHERBLOCK``
- ``FALLINGBLOCK``
- ``SCOREBOX``
- ``APOGEELOGO``

Otherwise, it was just the bitmaps, masked bitmaps, and ceilwalker sprites which needed changing.

(I've also added a hacky ``--igrab-underscore-separator`` option to my idGrab tool to automatically add an underscore separator before chunk suffixes. A similar hack could easily be made to IGRAB itself if needed.)

lethal-guitar commented:
> Sounds good — I've pushed a new version which adds the underscore consistently

Awesome, thanks! 👍🏻 I had to adjust two occurences of `PLATFORMSPR` and `APOGEELOGOSPR` in `BM_ACT1.C`, but otherwise it builds fine and the resulting binary is still identical to the original one. I'll go ahead and merge this 🙂 

> A similar hack could easily be made to IGRAB itself if needed

Interesting, has the IGRAB code been released in the meantime? I seem to recall that it wasn't available, but might be misremembering

sulix commented:
> Interesting, has the IGRAB code been released in the meantime? I seem to recall that it wasn't available, but might be misremembering

It's in the [TED5 source](https://legacy.3drealms.com/stuff/ted5.zip), under ``TEDTILES/IGRABSRC.ZIP``.
